nrf-hal-common/src/ieee802154.rs

```diff
@@ -305,22 +348,24 @@ impl<'c> Radio<'c> {
     /// validated by the hardware; otherwise it returns the `Err` variant. In either case, `packet`
     /// will be updated with the received packet's data
     pub fn recv(&mut self, packet: &mut Packet) -> Result<u16, u16> {
+        // Start non-blocking receive
+        let mut recv = self.recv_non_blocking(packet);
+
+        // Block untill receive is done
+        block!(recv.is_done())
+    }
+
+    /// Receives one radio packet and copies its contents into the given `packet` buffer
+    ///
+    /// This method is non-blocking
+    pub fn recv_non_blocking(&mut self, packet: &mut Packet) -> Recv<'_, 'c> {
         // Start the read
         // NOTE(unsafe) We block until reception completes or errors
```

Comment:
This is no longer true. `start_recv` doesn't have its safety conditions documented, but I suspect it requires the packet reference passed in to live at least until the receive completes. You don't enforce that here, so `recv_non_blocking` is unsound. You either need to make it unsafe (and document that the caller must ensure the `packet` lives at least until the receive operation completes), or add some safe abstraction to ensure that it does, such as having the `Radio` handle allocation of the packet.

Reply:
Good catch! I've added a mutable reference to the packet in `Recv` to enforce this.

Reply:
Unfortunately that's still not sound. The caller can call `forget` on the returned `Recv` and then drop the `Packet` that they passed a reference to, while the hardware still tries to write to it.

Reply:
If we require a pinned pointer for `Packet`, we get a drop guarantee.

Reply:
No, that won't help as we can still `forget` `Recv` and then `drop` the `Packet`. It's the `Recv` that needs to be dropped, not `Packet`. I guess unsafe is the best option then.

Reply:
Tried something new. `recv_non_blocking` takes a closure that takes `&Recv` as argument.

Reply:
Interesting approach, it's not a pattern I've seen much before for something like this but I think it's sound.

Reply:
If it's a soundness issue we might want to get a second review before we merge. I can try, but I'm bad at that stuff: might want to tag somebody?